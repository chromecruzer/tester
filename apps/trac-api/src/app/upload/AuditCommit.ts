import {UploadCommitA} from "./UploadCommitA";
import UploadStoreManagement from "../datastore/UploadStoreManagement";
import {datumToSql} from "@trac/postgresql";
import util from "util";
import _ from "lodash";
import {AuditMatch, AuditRecord, auditRecordFromJson} from "@trac/datatypes";
//need to keep this for errors
const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class AuditCommit extends UploadCommitA {
  private insertLocationSql: string;
  private insertLocationSqlSuffix: string;
  private insertItemsSql: string;
  private insertItemsSqlSuffix: string;
  private insertNotesSql: string;
  private record: AuditRecord;

  constructor(dataStore: UploadStoreManagement, schemaName: string,
              auditTableName: string, itemsTableName: string, notesTableName: string) {
    super(dataStore, schemaName);
    // console.log('constructor start');
    this.insertLocationSql = `INSERT INTO ${schemaName}.${auditTableName} VALUES (
gen_random_uuid(), `;
    this.insertLocationSqlSuffix = `) RETURNING uuid;`;
    this.insertItemsSql = `INSERT INTO ${schemaName}.${itemsTableName} VALUES `;
    this.insertItemsSqlSuffix = ` RETURNING lot, uuid;`;
    this.insertNotesSql = `INSERT INTO ${schemaName}.${notesTableName} VALUES `;
    // console.log('constructor stop');
  }

  public getArchiveData() {
    return this.record;
  }

  public async commit(token, client) {
    // console.log('commit start');
    this.record = auditRecordFromJson(await this.dataStore.retrieve(token));
    const location = await this.commitLocation(this.record.location, client);
    return this.commitItems(this.record.items, location, client)
      .then(result => {
        // console.log('commit stop');
        return result;
      });
  }

  private async commitLocation(location, client) {
    // console.log(`commit location ${dump(location)} ${datumToSql(location.scan_date, 'date')}`);
    const sql = `${this.insertLocationSql}
    ${datumToSql(location.received_date, 'date')},
    ${datumToSql(location.location_code, 'string')},
    ${datumToSql(location.scan_date, 'date')},
    null,
    ${datumToSql(location.status, 'string')},
    ${datumToSql(location.auditor, 'string')},
    ${datumToSql(location.scanner, 'string')}
    ${this.insertLocationSqlSuffix}`;
    return client.query(sql)
      .then(result => {
        location.uuid = result.rows[0].uuid;
        // console.log(`commit location stop rows ${dump(result.rows)} uuid ${result.rows[0].uuid}
        //  location ${dump(location)}`);
        return location;
      })
      .catch(err => {
        throw new Error(`Mistake in '${sql}' caused ${err.message}`);
      });
  }

  private async commitItems(items, location, client) {
    // console.log(`commit items start location ${dump(location)}`);
    const values = _.map(items, i => {
      return `(
    gen_random_uuid(),
    ${datumToSql(i.lot, 'string')},
    ${datumToSql(i.consignment_uuid, 'string')},
    ${datumToSql(i.consignment_location, 'string')},
    ${datumToSql(i.consignment_warehouse, 'string')},
    ${datumToSql(i.item, 'string')},
    ${datumToSql(i.family, 'string')},
    ${datumToSql(i.description, 'string')},
    ${datumToSql(i.expire_date, 'date')},
    ${datumToSql(i.audit_match, 'string')},
    ${datumToSql(location.uuid, 'string')},
    ${datumToSql(i.quantity, 'number')},
    ${datumToSql(i.last_changed_date, 'date')},
    ${datumToSql(i.confirmed_date, 'date')},
    ${datumToSql(i.confirmed_by, 'string')}
    )`;
    });
    const sql = `${this.insertItemsSql}
    ${_.join(values, ',\n')}
    ${this.insertItemsSqlSuffix}`;
    // console.log(`assembled sql ${sql}`);
    return client.query(sql)
      .then(result => {
        result.rows.forEach(r => {
          const found = _.find(items, i => i.lot === r.lot);
          if (!found) {
            return Promise.reject(new Error(`${dump(result)} not found in ${dump(items)}`));
          }
          found.uuid = r.uuid;
        });
        // console.log('commit items stop');
        return result;
      })
      .catch(err => {
        throw new Error(`Mistake in '${sql}' caused ${err.message}`);
      });
  }
}
