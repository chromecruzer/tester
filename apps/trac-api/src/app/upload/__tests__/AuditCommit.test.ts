import {DateTime} from "luxon";
import AuditCommit from "../AuditCommit";

describe('AuditCommit', () => {
  let storageMock, clientMock, auditRecord, committer;
  const now = DateTime.fromObject({year: 2022, month: 6, day: 15}).toJSDate();
  // console.log(`now for testing ${now} ${now}`);
  beforeEach(() => {
    auditRecord = {
      location: {
        location_code: 'I am a location code also known as a customer id or a customer code',
        scan_date: DateTime.fromObject({year: 2022, month: 2, day: 15}).toJSDate(),
        status: 'Open',
        auditor: 'I am the auditor',
        scanner: 'I am the scanner',
      },
      items: [
        {
          lot: 'I am a lot',
          consignment_uuid: 'consignment_uuid1',
          family: 'falcon crest',
          description: 'mister mister',
          item: 'i am an item',
          quantity: 1,
          expire_date: DateTime.fromObject({year: 2023, month: 1, day: 15}).toJSDate(),
          audit_match: 'True Match',
          last_changed_date: now,
        },
        {
          lot: 'I am also a lot',
          consignment_uuid: 'consignment_uuid2',
          consignment_location: 'consignment location',
          family: 'falcon crest',
          description: 'mister mister',
          item: 'i am an item',
          quantity: 1,
          expire_date: DateTime.fromObject({year: 2023, month: 1, day: 15}).toJSDate(),
          audit_match: 'Found In Other Location',
          last_changed_date: now,
        },
        {
          lot: 'I am a lot with no consignment',
          audit_match: 'No Match',
          last_changed_date: now,
        },
        {
          lot: 'I am a lot that is not in the scan',
          consignment_uuid: 'consignment_uuid2',
          family: 'falcon crest',
          description: 'mister mister',
          item: 'i am an item',
          quantity: 1,
          expire_date: DateTime.fromObject({year: 2023, month: 1, day: 15}).toJSDate(),
          audit_match: 'Missing',
          last_changed_date: now,
        },
      ]
    };
    storageMock = {
      retrieve: jest.fn().mockReturnValue(auditRecord)
    };
    clientMock = {
      query: jest.fn()
    }
    committer = new AuditCommit(storageMock, ' schema name', 'audit location table',
      'audit item table', 'audit notes table')
  });
  describe('commit', () => {
    it('should correctly store an audit record', () => {
      clientMock.query.mockReturnValueOnce(Promise.resolve({rows: [{uuid: 'location uuid'}]}));
      clientMock.query.mockReturnValueOnce(Promise.resolve({rows: [
          {
            lot: 'I am a lot',
            uuid: 'uuid1'
          },
          {
            lot: 'I am also a lot',
            uuid: 'uuid2'
          },
          {
            lot: 'I am a lot with no consignment',
            uuid: 'uuid3'
          },
          {
            lot: 'I am a lot that is not in the scan',
            uuid: 'uuid4'
          },
        ]}));
      clientMock.query.mockReturnValueOnce(Promise.resolve());
      committer.commit('I am a token', clientMock)
        .then(() => {
          // console.log('commit resolved')
          expect(clientMock.query.mock.calls).toMatchSnapshot();
        })
    });
  });
});
