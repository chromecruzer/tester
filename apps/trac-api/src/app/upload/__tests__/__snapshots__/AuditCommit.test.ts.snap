// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuditCommit commit should correctly store an audit record 1`] = `
Array [
  Array [
    "INSERT INTO  schema name.audit location table VALUES (
gen_random_uuid(), 
    null,
    'I am a location code also known as a customer id or a customer code',
    '02/15/2022',
    null,
    'Open',
    'I am the auditor',
    'I am the scanner'
    ) RETURNING uuid;",
  ],
  Array [
    "INSERT INTO  schema name.audit item table VALUES 
    (
    gen_random_uuid(),
    'I am a lot',
    'consignment_uuid1',
    null,
    'i am an item',
    'falcon crest',
    'mister mister',
    '01/15/2023',
    'True Match',
    'location uuid',
    1,
    '06/15/2022',
    null,
    null
    ),
(
    gen_random_uuid(),
    'I am also a lot',
    'consignment_uuid2',
    'consignment location',
    'i am an item',
    'falcon crest',
    'mister mister',
    '01/15/2023',
    'Found In Other Location',
    'location uuid',
    1,
    '06/15/2022',
    null,
    null
    ),
(
    gen_random_uuid(),
    'I am a lot with no consignment',
    null,
    null,
    null,
    null,
    null,
    null,
    'No Match',
    'location uuid',
    null,
    '06/15/2022',
    null,
    null
    ),
(
    gen_random_uuid(),
    'I am a lot that is not in the scan',
    'consignment_uuid2',
    null,
    'i am an item',
    'falcon crest',
    'mister mister',
    '01/15/2023',
    'Missing',
    'location uuid',
    1,
    '06/15/2022',
    null,
    null
    )
     RETURNING lot, uuid;",
  ],
]
`;
